### PR TITLE
Mgmt 2230 added authorization scopes (with read only admin)

### PR DESCRIPTION
**Pull request by RazRegev**

### CODE REVIEW

File: pkg/auth/auth_handler.go [hunk 1]
```diff
@@ -50,6 +55,12 @@ func NewAuthHandler(cfg Config, ocmCLient *ocm.Client, log logrus.FieldLogger) *
 	return a
 }
 
+func getAdminUsers(cfg Config, log logrus.FieldLogger) []string {
+	admins := strings.ReplaceAll(cfg.AdminUsers, " ", "")
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:59]: I think that this is not necessary.
all you need to do is set `AdminUsers` as `[]string` in the config structure and envconfig should do the parsing

File: pkg/auth/auth_handler.go [hunk 2]
```diff
@@ -193,16 +205,20 @@ func (a *AuthHandler) AuthUserAuth(token string) (interface{}, error) {
...
+
+func (a *AuthHandler) storeReadOnlyAdminInPayload(payload *ocm.AuthPayload) error {
+	readOnlyAdmin, err := a.isReadOnlyAdmin(payload.Username)
 	if err != nil {
 		return err
 	}
-	payload.IsAdmin = admin
+	payload.IsReadOnlyAdmin = readOnlyAdmin
 	return nil
 }
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:220]: I think that payload should not have both `isReadOnlyAdmin` and `isAdmin`
We cannot have both of them at the same time

File: pkg/ocm/client.go [hunk 3]
```diff
@@ -117,13 +117,15 @@ func (c *Client) newConnection() error {
...
+	Username        string   `json:"username"`
+	FirstName       string   `json:"first_name"`
+	LastName        string   `json:"last_name"`
+	Organization    string   `json:"org_id"`
+	Email           string   `json:"email"`
+	Issuer          string   `json:"iss"`
+	ClientID        string   `json:"clientId"`
+	IsAdmin         bool     `json:"is_admin"`
+	IsReadOnlyAdmin bool     `json:"is_read_only_admin"`
+	IsUser          bool     `json:"is_user"`
```
Comment (ronniel1) [inline pkg/ocm/client.go:129]: doesn't make sense to have three of them since user can be one or the other.
Maybe this should be en enum?

File: swagger.yaml [hunk 4]
```diff
@@ -35,6 +35,8 @@ paths:
     post:
       tags:
         - installer
+      security:
+        - userAuth: [admin, user]
```
Comment (ronniel1) [inline swagger.yaml:39]: Since most APIs are permitted for `[read-only-admin, user]` I suggest that this be the default and for the others define it per API
The default is defined in line 31 above

File: Makefile [hunk 5]
```diff
@@ -43,6 +43,10 @@ ifdef INSTALLATION_TIMEOUT
         INSTALLATION_TIMEOUT_FLAG = --installation-timeout $(INSTALLATION_TIMEOUT)
 endif
 
+ifdef ADMIN_USERS
```
Comment (ronniel1) [inline Makefile:46]: I don't think we need this here
For cloud deployments we'll get these values from the saas.yml
for non-cloud deployments we are using no-auth and then all users ar admin

@filanov , @YuviGold  what do you think?

Comment (filanov) [reply Makefile:46]: Agree

File: pkg/auth/authz_handler.go [hunk 6]
```diff
@@ -105,3 +119,44 @@ func OrgIDFromContext(ctx context.Context) string {
 	payload := PayloadFromContext(ctx)
 	return payload.Organization
 }
+
+func (a *AuthzHandler) isAuthorizedToAccessRoute(
+	request *http.Request,
+	payload *ocm.AuthPayload) bool {
+
+	route := middleware.MatchedRouteFrom(request)
+	if route.Authenticators == nil {
```
[the remark this answers is not in the document]
Comment (RazRegev) [reply pkg/auth/authz_handler.go:126]: yes

File: pkg/auth/authz_handler.go [hunk 7]
```diff
@@ -105,3 +119,44 @@ func OrgIDFromContext(ctx context.Context) string {
...
+	}
+
+	authScheme := route.Authenticator.Schemes[0]
+	for _, policy := range route.Authenticators {
+		policyScopes, exist := policy.Scopes[authScheme]
+		if !exist {
+			continue
+		}
+		if len(policyScopes) == 0 {
+			a.log.Debugf(
```
[the remark this answers is not in the document]
Comment (RazRegev) [reply pkg/auth/authz_handler.go:140]: not sure I understand. if I get it right I need to iterate over the policies twice - one time for the isPolicyScopesAvailable function, and other time here when comparing scopes. is this what you meant?

File: pkg/ocm/client.go [hunk 8]
```diff
@@ -115,15 +115,24 @@ func (c *Client) newConnection() error {
...
-	IsUser       bool   `json:"is_user"`
+	Username     string   `json:"username"`
+	FirstName    string   `json:"first_name"`
+	LastName     string   `json:"last_name"`
+	Organization string   `json:"org_id"`
+	Email        string   `json:"email"`
+	Issuer       string   `json:"iss"`
+	ClientID     string   `json:"clientId"`
+	Role         RoleType `json:"role"`
+	Scopes       []string `json:"scopes"`
```
Comment (ronniel1) [inline pkg/ocm/client.go:137]: No need for scopes
Role should be either RO-admin or admin or user

File: pkg/auth/authz_handler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielerez) [reply pkg/auth/authz_handler.go:124]: Yes, that's why I wondered if we really need the loop, as we only use a single authenticator.
Maybe extracting only the logging to a separate func would make it a bit clearer (just to avoid duplication and make the func shorter)

[the remark this answers is not in the document]
Comment (RazRegev) [reply pkg/auth/authz_handler.go]: done

File: pkg/ocm/client.go [hunk 9]
```diff
@@ -121,6 +121,15 @@ func (c *Client) newConnection() error {
 	return nil
 }
 
+type Scope string
```
Comment (ronniel1) [inline pkg/ocm/client.go:124]: Scope -> RoleType

Suggest to add the "type" fo avoid confusing with "Role" property in line 142 below

File: pkg/ocm/client.go [hunk 10]
```diff
@@ -130,6 +139,5 @@ type AuthPayload struct {
 	Email        string `json:"email"`
 	Issuer       string `json:"iss"`
 	ClientID     string `json:"clientId"`
-	IsAdmin      bool   `json:"is_admin"`
-	IsUser       bool   `json:"is_user"`
+	Scope        Scope  `json:"scope"`
```
Comment (ronniel1) [inline pkg/ocm/client.go:142]: Scope -> Role

File: pkg/auth/auth_handler.go [hunk 11]
```diff
@@ -194,25 +199,30 @@ func (a *AuthHandler) AuthUserAuth(token string) (interface{}, error) {
...
 
-func (a *AuthHandler) isAdmin(username string) (bool, error) {
+func (a AuthHandler) getScope(payload *ocm.AuthPayload) (ocm.Scope, error) {
+	if funk.Contains(a.AdminUsers, payload.Username) {
+		return ocm.AdminScope, nil
+	}
+	isReadOnly, err := a.isReadOnlyAdmin(payload.Username)
+	if err != nil {
+		return ocm.UndefinedScope, err
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:217]: I think default here should be `user`, no?

File: pkg/auth/auth_handler.go [hunk 12]
```diff
@@ -194,25 +199,30 @@ func (a *AuthHandler) AuthUserAuth(token string) (interface{}, error) {
...
 	}
-	payload.IsAdmin = admin
-	a.client.Cache.Set(payloadKey, payload, cache.DefaultExpiration)
-
+	payload.Scope = scope
 	return nil
 }
 
-func (a *AuthHandler) isAdmin(username string) (bool, error) {
+func (a AuthHandler) getScope(payload *ocm.AuthPayload) (ocm.Scope, error) {
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:211]: getScope -> getRole

File: pkg/auth/authz_handler.go [hunk 13]
```diff
@@ -51,34 +53,28 @@ func (a *AuthzHandler) CreateAuthorizer() func(*http.Request) error {
 // and the principal was stored in the context in the "AuthKey" context value.
 func (a *AuthzHandler) Authorizer(request *http.Request) error {
 	payload := PayloadFromContext(request.Context())
-	username := payload.Username
-	payloadFromCache, found := a.client.Cache.Get(username)
```
Comment (ronniel1) [inline pkg/auth/authz_handler.go:55]: Why did you remove the cache?
The cache should only be on the allowedToUseAssistedInstaller() call

File: pkg/auth/authz_handler.go [hunk 14]
```diff
@@ -105,3 +101,44 @@ func OrgIDFromContext(ctx context.Context) string {
...
+	}
+
+	authScheme := route.Authenticator.Schemes[0]
+	for _, policy := range route.Authenticators {
+		policyScopes, exist := policy.Scopes[authScheme]
+		if !exist {
+			continue
+		}
+		if len(policyScopes) == 0 {
+			return allScopesAreAllowedResponse()
```
Comment (ronniel1) [inline pkg/auth/authz_handler.go:129]: don't we need to also `continue` in this case?

File: pkg/auth/authz_handler.go [hunk 15]
```diff
@@ -51,34 +53,28 @@ func (a *AuthzHandler) CreateAuthorizer() func(*http.Request) error {
...
+			http.StatusUnauthorized,
+			fmt.Errorf(
+				"%s: Unauthorized to access route (access review failed)",
+				payload.Username))
 	}
-
-	if payload.IsUser {
-		// authorized user
-		return nil
+	isAuthorized = a.isAuthorizedToAccessRoute(request, payload)
```
Comment (ronniel1) [inline pkg/auth/authz_handler.go:67]: This should probably be done before calling allowedToUseAssistedInstaller()
since allowedToUseAssistedInstaller involves calling a different service

File: pkg/ocm/client.go (hunk 9 above)
Comment (RazRegev) [reply pkg/ocm/client.go:124]: This is how I originally started but I changed role to scope so it won't be confusing when comparing "role" to "scope" in authorization context.

File: pkg/auth/authz_handler.go (hunk 14 above)
Comment (RazRegev) [reply pkg/auth/authz_handler.go:129]: If we get to this line it means the route is defined with an empty list for the required auth.
Currently it can only happen only with agentAuth, in such cases all scopes are allowed.

File: pkg/auth/auth_handler.go [hunk 16]
```diff
@@ -194,25 +199,30 @@ func (a *AuthHandler) AuthUserAuth(token string) (interface{}, error) {
 	return payload, nil
 }
 
-func (a *AuthHandler) storeAdminInPayload(payload *ocm.AuthPayload) error {
-	payloadKey := payload.Username + "_is_admin"
-	payloadFromCache, found := a.client.Cache.Get(payloadKey)
-	if found {
```
[the remark this answers is not in the document]
Comment (RazRegev) [reply pkg/auth/auth_handler.go:200]: done

File: pkg/auth/authz_handler.go (hunk 15 above)
Comment (RazRegev) [reply pkg/auth/authz_handler.go]: good point I should have noticed that

File: pkg/auth/auth_handler.go [hunk 17]
```diff
@@ -185,34 +190,50 @@ func (a *AuthHandler) AuthUserAuth(token string) (interface{}, error) {
...
 
-func (a *AuthHandler) storeAdminInPayload(payload *ocm.AuthPayload) error {
+func (a AuthHandler) storeRoleInUserAuthPayload(payload *ocm.AuthPayload) error {
 	payloadKey := payload.Username + "_is_admin"
-	payloadFromCache, found := a.client.Cache.Get(payloadKey)
-	if found {
-		payload.IsAdmin = payloadFromCache.(*ocm.AuthPayload).IsAdmin
-		return nil
+	if payloadFromCache, existInCache := a.client.Cache.Get(payloadKey); existInCache {
+		payload.Role = payloadFromCache.(*ocm.AuthPayload).Role
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:203]: Cache handling to not be here, but in AuthUserAuth()

Also you need to add to the cache if user it was not in cache and error < 500

File: pkg/auth/auth_handler.go [hunk 18]
```diff
@@ -93,7 +98,12 @@ func (a *AuthHandler) AuthAgentAuth(token string) (interface{}, error) {
 		a.log.Errorf("Error Authenticating PullSecret token: %v", err)
 		return nil, common.ApiErrorWithDefaultInfraError(err, http.StatusUnauthorized)
 	}
-	err = a.storeAdminInPayload(user)
+	err = a.storeRoleInPayload(user)
+
+	if shouldStorePayloadInCache(err) {
+		a.client.Cache.Set(token, user, cache.DefaultExpiration)
```
Comment (ronniel1) [inline pkg/auth/auth_handler.go:104]: Line 111 also stores the same in cache

File: pkg/auth/auth_utils.go [hunk 19]
```diff
@@ -23,3 +26,15 @@ func AgentAuthHeaderWriter(token string) runtime.ClientAuthInfoWriter {
 func UserAuthHeaderWriter(token string) runtime.ClientAuthInfoWriter {
 	return AuthHeaderWriter(token, userAuthHeader)
 }
+
+func shouldStorePayloadInCache(err error) bool {
+	if err == nil {
+		return true
+	}
+	if serr, ok := err.(*common.ApiErrorResponse); ok {
```
Comment (ronniel1) [inline pkg/auth/auth_utils.go:34]: this function should return true if error is  less then 500

File: tools/deploy_assisted_installer_configmap.py [hunk 20]
```diff
@@ -97,5 +101,14 @@ def main():
     )
 
 
+def get_admin_users():
+    admins_file = os.path.join(os.getcwd(), 'ADMINS')
```
Comment (YuviGold) [inline tools/deploy_assisted_installer_configmap.py:105]: Can we set the admins on [OWNERS_ALIASES](https://github.com/openshift/assisted-service/blob/master/OWNERS_ALIASES) ?

Comment (ronniel1) [reply tools/deploy_assisted_installer_configmap.py:105]: I don't think we should combine these

Comment (YuviGold) [reply tools/deploy_assisted_installer_configmap.py:105]: I thought these are the same usernames. ignore

**Review by YuviGold — APPROVED** [review]
